src/utils.ts

```diff
@@ -26,3 +32,91 @@ export function toWei(
 export function fromWei(amount: BigNumber, decimals: number): BigNumber {
   return amount.dividedBy(TEN.pow(decimals));
 }
+
+export type SummaryEntry = {
+  tokenAddress: string;
+  amount: BigNumber;
+  decimals: number;
+};
+
+export const transfersToSummary = (transfers: Payment[]) => {
+  return transfers.reduce((previousValue, currentValue): Map<
+    string,
+    SummaryEntry
+  > => {
+    let tokenSummary = previousValue.get(currentValue.tokenAddress);
+    if (typeof tokenSummary === "undefined") {
+      tokenSummary = {
+        tokenAddress: currentValue.tokenAddress,
+        amount: new BigNumber(0),
+        decimals: currentValue.decimals,
+      };
+      previousValue.set(currentValue.tokenAddress, tokenSummary);
+    }
+    tokenSummary.amount = tokenSummary.amount.plus(currentValue.amount);
+
+    return previousValue;
+  }, new Map<string, SummaryEntry>());
+};
+
+export type InsufficientBalanceInfo = {
+  token: string;
+  transferAmount: string;
+};
+
+export const checkAllBalances = async (
+  summary: Map<string, SummaryEntry>,
+  web3Provider: ethers.providers.Web3Provider,
+  safe: SafeInfo,
+  tokenList: TokenMap
+): Promise<InsufficientBalanceInfo[]> => {
+  const insufficientTokens: InsufficientBalanceInfo[] = [];
+  for (const summaryEntry of summary.values()) {
+    const tokenAddress = summaryEntry.tokenAddress;
+    const tokenAmount = summaryEntry.amount;
+    if (tokenAddress === null) {
+      // Check ETH Balance
+      const tokenBalance = await web3Provider.getBalance(
+        safe.safeAddress,
+        "latest"
+      );
+      if (!isSufficientBalance(tokenBalance, tokenAmount, 18)) {
+        insufficientTokens.push({
+          token: "ETH",
+          transferAmount: tokenAmount.toFixed(),
+        });
+      }
+    } else {
+      const erc20Contract = erc20Instance(
```

Comment:
Don't we need to await this contract instance? I believe `erc20Instance` should be async...

Reply:
I see we await the balance later. However, I wonder what would happen here is there is no contract at that address.

Reply:
I think the Contract Object itself is not async. But all the calls on it are.

I will see what happens if I just put a random address not belonging to a token:
My guess is either the balance will just return 0x0 and thus be too low or an error occurs because its not a contract.